src/parser/statement.js

```diff
@@ -976,54 +965,86 @@ export default class StatementParser extends ExpressionParser {
     isStatic: boolean,
   ) {
     const memberAny: any = member;
-    const methodOrProp: N.ClassMethod | N.ClassProperty = memberAny;
-    const method: N.ClassMethod = memberAny;
-    const prop: N.ClassProperty = memberAny;
+    const methodOrProp:
+      | N.ClassMethod
+      | N.ClassPrivateMethod
+      | N.ClassProperty
+      | N.ClassPrivateProperty = memberAny;
+    const method: N.ClassMethod | N.ClassPrivateMethod = memberAny;
+    const prop: N.ClassProperty | N.ClassPrivateProperty = memberAny;
 
     methodOrProp.static = isStatic;
 
     if (this.eat(tt.star)) {
       // a generator
       method.kind = "method";
-      this.parsePropertyName(method);
+      this.parseClassPropertyName(method);
+
+      if (method.key.type === "PrivateName") {
+        this.expectPlugin("classPrivateMethods", method.key.start);
+        // Private generator method
+        this.parseClassPrivateMethod(
+          classBody,
+          method,
+          true,
+          false,
+          /* isConstructor */ false,
+        );
+        return;
+      }
+
       if (this.isNonstaticConstructor(method)) {
         this.raise(method.key.start, "Constructor can't be a generator");
       }
-      if (
-        !method.computed &&
-        method.static &&
-        (method.key.name === "prototype" || method.key.value === "prototype")
-      ) {
-        this.raise(
-          method.key.start,
-          "Classes may not have static property named prototype",
-        );
-      }
+
       this.parseClassMethod(
         classBody,
         method,
         true,
         false,
         /* isConstructor */ false,
       );
+
       return;
     }
 
-    const isSimple = this.match(tt.name);
     const key = this.parseClassPropertyName(methodOrProp);
+    let isPrivate = key.type === "PrivateName";
+    const isSimple = !isPrivate && !key.computed;
+
+    const handleClassProperty = () => {
+      if (isPrivate) {
+        this.expectPlugin("classPrivateProperties", key.start);
+        // Private property
+        classBody.body.push(this.parseClassPrivateProperty(prop));
```

Comment:
Can we make a `#pushClassPrivateProperty`?

Reply:
Done -- should the two be merged or remain separate?